lua/packer.lua

```diff
@@ -663,71 +663,84 @@ end
-          configs_to_run[plugin_name] = plugin_info.config
+  async(function()
+    if not donot_move_plugins then
+      local update = require_and_configure 'update'
+      local plugin_utils = require_and_configure 'plugin_utils'
+      local fs_state = await(plugin_utils.get_fs_state(plugins))
+      await(a.main)
+      update.fix_plugin_types(plugins, vim.tbl_keys(fs_state.missing), {}, fs_state)
+    end
+    local args = parse_args(raw_args)
+    local output_path = args.output_path or config.compile_path
+    local output_lua = vim.fn.fnamemodify(output_path, ':e') == 'lua'
+    local should_profile = args.profile
+    -- the user might explicitly choose for this value to be false in which case
+    -- an or operator will not work
+    if should_profile == nil then
+      should_profile = config.profile.enable
+    end
+    refresh_configs(plugins)
+    -- NOTE: we copy the plugins table so the in memory value is not mutated during compilation
+    local compiled_loader = compile(vim.deepcopy(plugins), output_lua, should_profile)
+    output_path = vim.fn.expand(output_path)
+    vim.fn.mkdir(vim.fn.fnamemodify(output_path, ':h'), 'p')
+    local output_file = io.open(output_path, 'w')
+    output_file:write(compiled_loader)
+    output_file:close()
+    if config.auto_reload_compiled then
+      local configs_to_run = {}
+      if _G.packer_plugins ~= nil then
+        for plugin_name, plugin_info in pairs(_G.packer_plugins) do
+          if plugin_info.loaded and plugin_info.config and plugins[plugin_name] and plugins[plugin_name].cmd then
+            configs_to_run[plugin_name] = plugin_info.config
+          end
         end
       end
-    end
 
-    vim.cmd('source ' .. output_path)
-    for plugin_name, plugin_config in pairs(configs_to_run) do
-      for _, config_line in ipairs(plugin_config) do
-        local success, err = pcall(loadstring(config_line))
-        if not success then
-          vim.notify('Error running config for ' .. plugin_name .. ': ' .. vim.inspect(err), vim.log.levels.ERROR, {})
+      vim.cmd('source ' .. output_path)
+      for plugin_name, plugin_config in pairs(configs_to_run) do
+        for _, config_line in ipairs(plugin_config) do
+          local success, err = pcall(loadstring(config_line))
+          if not success then
+            vim.notify('Error running config for ' .. plugin_name .. ': ' .. vim.inspect(err), vim.log.levels.ERROR, {})
+          end
         end
       end
     end
-  end
-  log.info 'Finished compiling lazy-loaders!'
-  packer.on_compile_done()
-
-  -- TODO: remove this after migration period (written 2021/06/28)
-  local old_output_path
-  if output_lua then
-    old_output_path = vim.fn.fnamemodify(output_path, ':r') .. '.vim'
-  else
-    old_output_path = vim.fn.fnamemodify(output_path, ':r') .. '.lua'
-  end
+    log.info 'Finished compiling lazy-loaders!'
+    print 'PackerCompile: Done'
```

Comment:
We should remove this print statement before merging.

Reply:
This isn't for debugging but to let user know compilation is done . Since compilation is async now . :PackerCompile exits immediately and there's no way to know when it's done . I'm open for better way to let user know but there should be a way at lest :)

Reply:
@shadmansaleh how about using `vim.notify`?

Reply:
> how about using vim.notify?

Works for me.

Reply:
Is there a similar thing for if compilation errors? this has always been a thing that I've hit quite often historically using packer and I currently have the call to `packer.sync` wrapped in a `pcall` to avoid explosions from typos or syntax errors, haven't properly reviewed this but is there a point where the errors can be reported to the user as well or how will that come through now?

Reply:
Oh nevermind, there's already vim.notify calls for errors

Reply:
You'll be getting messages about errors . But yeah you can't catch them with pcall 😅

Can you catch errors in `packer.sync` ? it's aleady async isn't it ?

Reply:
Sorry that was a typo I meant `compile`

Reply:
> Oh nevermind, there's already vim.notify calls for errors

Ohh right . Now I remember had added it for functions in #343 . The handling errors in sting is still hanging #392 😐